synapse/rest/media/v1/media_repository.py

```diff
@@ -539,7 +539,10 @@ def _generate_thumbnails(self, server_name, media_id, media_info, url_cache=Fals
             Deferred[dict]: Dict with "width" and "height" keys of original image
         """
         media_type = media_info["media_type"]
-        file_id = media_info.get("filesystem_id")
+        if server_name:
+            file_id = media_info["filesystem_id"]
+        else:
+            file_id = media_id
```

Comment:
the fact that `media_info["filesystem_id"]` and `media_id` give different values feels like a red flag to me. why doesn't  `media_info["filesystem_id"]` work for the local case?

Reply:
Mainly because we handle local and remote media differently in the DB, and so it returns different things. I think what has happened is that the local/remote code paths were almost entirely separate and have slowly converged over time, making this a bit strange.

Reply:
Ok. I've now had a bit of a look at how this works. Taking a `dict` which can take one of two completely unspecified shapes is a horrible API. `media_info` is only used to pass the media_type and the file_id. Why don't we make both of those explicit parameters?

Reply:
[Note to self: this is effectively reinstating the condition at https://github.com/matrix-org/synapse/pull/2767/files#diff-043e4890d840f3862222c34adf3a9459L476, which constructed the local filename from either file_id or media_id. We now always use file_id so need to get it right]